File: hacking_with_python/insecure_authentication.py
import requests
from colorama import init, Fore, Style

# Initialize colorama to enable colored output on Windows
init()

def test_insecure_authentication_vulnerability(target_url):
    print(Fore.YELLOW + "Testing for Insecure Authentication vulnerability on:", target_url)
    try:
        # Sending a GET request without authentication
        response = requests.get(target_url)
        
        # Checking the status code for unauthorized access (401)
        if response.status_code == 401:
            print(Fore.RED + "Insecure Authentication vulnerability found!")
            print(Fore.GREEN + "Details:")
            print("The server responded with a 401 Unauthorized status code, indicating that access is restricted.")
            print("This vulnerability allows unauthorized users to access restricted pages or endpoints without proper authentication.")
            print(Fore.GREEN + "Recommendations:")
            print("1. Implement proper authentication mechanisms such as username/password or token-based authentication.")
            print("2. Ensure that all sensitive resources are protected and require valid authentication credentials.")
            print("3. Regularly audit and review access controls to identify and mitigate potential vulnerabilities.")
        else:
            print(Fore.GREEN + "No Insecure Authentication vulnerability found.")
    except requests.exceptions.RequestException as e:
        print(Fore.RED + "Error occurred:", e)

if __name__ == "__main__":
    # Asking for the target website URL
    target_url = input(Fore.CYAN + "Enter your target website url: ")

    # Testing for Insecure Authentication vulnerability
    test_insecure_authentication_vulnerability(target_url)

File: hacking_with_ruby/insecure_authentication.rb
require 'net/http'
require 'colorize'

def test_insecure_authentication_vulnerability(url)
  uri = URI(url)
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true if uri.scheme == 'https'

  request = Net::HTTP::Get.new(uri.request_uri)
  
  begin
    response = http.request(request)
  
    if response.code == '200'
      puts "Insecure Authentication Vulnerability Detected!".colorize(:red)
      puts "The website is using HTTP Basic Authentication without encryption.".colorize(:red)
      puts "This can lead to credentials being transmitted over the network in plaintext.".colorize(:red)
      puts "To verify manually, try accessing the URL without HTTPS (e.g., http://#{uri.host}:#{uri.port}) and check if authentication is required.".colorize(:yellow)
    else
      puts "No Insecure Authentication Vulnerability Detected.".colorize(:green)
    end
  rescue StandardError => e
    puts "An error occurred while testing the vulnerability: #{e.message}".colorize(:yellow)
    puts "Please ensure that the target URL is accessible and try again.".colorize(:yellow)
  end
end

puts "Enter your target website URL: "
target_url = gets.chomp

test_insecure_authentication_vulnerability(target_url)
